ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/encoding/DepositLogDecoder.java

```diff
@@ -14,27 +14,92 @@
+ */
 public class DepositLogDecoder {
 
+  private static final int DEPOSIT_LOG_LENGTH = 576;
+  private static final int LENGTH_FIELD_SIZE = 32;
+
+  private static final int PUB_KEY_LENGTH = 48;
+  private static final int WITHDRAWAL_CRED_LENGTH = 32;
+  private static final int AMOUNT_LENGTH = 8;
+  private static final int SIGNATURE_LENGTH = 96;
+  private static final int INDEX_LENGTH = 8;
+
+  private static final int PUB_KEY_OFFSET = 0;
+  // PublicKey is the first element. ABI encoding pads values to 32 bytes, so
+  // despite BLS public keys being length 48, the value length here is 64. Then
+  // skip over the next length value.
+  private static final int WITHDRAWAL_CRED_OFFSET =
+      PUB_KEY_OFFSET + PUB_KEY_LENGTH + 16 + LENGTH_FIELD_SIZE;
+  // WithdrawalCredentials is 32 bytes. Read that value then skip over next length.
+  private static final int AMOUNT_OFFSET =
+      WITHDRAWAL_CRED_OFFSET + WITHDRAWAL_CRED_LENGTH + LENGTH_FIELD_SIZE;
+  // amount is only 8 bytes long but is stored in a 32 byte slot. The remaining 24 bytes need to be
+  // added to the offset.
+  private static final int SIGNATURE_OFFSET =
+      AMOUNT_OFFSET + AMOUNT_LENGTH + 24 + LENGTH_FIELD_SIZE;
+  // Signature is 96 bytes. Skip over it and the next length.
+  private static final int INDEX_OFFSET = SIGNATURE_OFFSET + SIGNATURE_LENGTH + LENGTH_FIELD_SIZE;
+
+  // The ABI encodes the position of dynamic elements first. Since there are 5
+  // elements, skip over the positional data. The first 32 bytes of dynamic
+  // elements also encode their actual length. Skip over that value too.
+  private static final int DATA_START_POSITION = 32 * 5 + LENGTH_FIELD_SIZE;
+
   public static Bytes decodeFromLog(final Log log) {
-    Contract.EventValuesWithLog eventValues = DepositContract.staticExtractDepositEventWithLog(log);
-    final Bytes rawPublicKey =
-        Bytes.wrap((byte[]) eventValues.getNonIndexedValues().get(0).getValue());
-    final Bytes rawWithdrawalCredential =
-        Bytes.wrap((byte[]) eventValues.getNonIndexedValues().get(1).getValue());
-    final Bytes rawAmount =
-        Bytes.wrap((byte[]) eventValues.getNonIndexedValues().get(2).getValue());
-    final Bytes rawSignature =
-        Bytes.wrap((byte[]) eventValues.getNonIndexedValues().get(3).getValue());
-    final Bytes rawIndex = Bytes.wrap((byte[]) eventValues.getNonIndexedValues().get(4).getValue());
-
-    return Bytes.concatenate(
-        rawPublicKey, rawWithdrawalCredential, rawAmount, rawSignature, rawIndex);
+    final Bytes data = log.getData();
+
+    if (data.size() != DEPOSIT_LOG_LENGTH) {
+      throw new IllegalArgumentException(
+          "Invalid deposit log length. Must be "
+              + DEPOSIT_LOG_LENGTH
+              + " bytes, but is "
+              + data.size()
+              + " bytes");
+    }
+
+    final Bytes depositData = data.slice(DATA_START_POSITION);
```

Comment:
You could get rid of this slice if you were to just add the DATA_START_POSITION to each of the offsets.

Reply:
Good idea, I've done that.